address/address.go

```diff
@@ -47,327 +77,236 @@ const (
-	sum := createChecksum(hrp, combined)
-
-	toEncode := append(combined, sum...)
-	encoded := make([]byte, len(toEncode))
-
-	for i, p := range toEncode {
-		if p >= 32 {
-			return "", fmt.Errorf("invalid data")
-		}
-		encoded[i] += Base32Charset[p]
+	var strAddr string
+	switch addr.Protocol() {
+	case SECP256K1, Actor, BLS:
+		cksm := Checksum(append([]byte{addr.Protocol()}, addr.Payload()...))
+		strAddr = ntwk + fmt.Sprintf("%d", addr.Protocol()) + AddressEncoding.WithPadding(-1).EncodeToString(append(addr.Payload(), cksm[:]...))
+	case ID:
+		strAddr = ntwk + fmt.Sprintf("%d", addr.Protocol()) + fmt.Sprintf("%d", leb128.ToUInt64(addr.Payload()))
+	default:
+		return "", errors.New("invalid protocol byte")
 	}
-
-	return hrp + string(encoded), nil
+	log.Debugf("encoded address: %s", strAddr)
+	return strAddr, nil
 }
 
-// decode decodes a filecoin address and returns
-// the hrp(human-readable part), version (1byte) and data(32bit data array)
-// or an error.
-func decode(addr string) (string, byte, []byte, error) {
-	if len(addr) < 2 {
-		return "", 0, nil, fmt.Errorf("address too short")
-	}
-	if len(addr) > 41 {
-		return "", 0, nil, fmt.Errorf("too long: len=%d", len(addr))
+func decode(a string) (Address, error) {
+	if len(a) == 0 {
+		return Undef, nil
 	}
-	if strings.ToLower(addr) != addr && strings.ToUpper(addr) != addr {
-		return "", 0, nil, fmt.Errorf("mixed case")
+	log.Debugf("decoding address: %s", a)
+	if len(a) < 3 {
+		return Undef, errors.Errorf("invalid address: %s length: %d, too short", a, len(a))
 	}
-	addr = strings.ToLower(addr)
-	hrp := addr[0:2]
-	for p, c := range hrp {
-		if c < 33 || c > 126 {
-			return "", 0, nil, fmt.Errorf("invalid character human-readable part: addr[%d]=%d", p, c)
-		}
-	}
-
-	// check data
-	data := addr[2:]
-	dataBytes := []byte{}
-	for _, b := range data {
-		// alphanumeric only
-		if !((b >= int32('0') && b <= int32('9')) || (b >= int32('A') && b <= int32('Z')) || (b >= int32('a') && b <= int32('z'))) {
-			return "", 0, nil, fmt.Errorf("non alphanumeric character")
-		}
 
-		// exclude [1,b,i,o]
-		if b == int32('1') || b == int32('b') || b == int32('i') || b == int32('o') {
-			return "", 0, nil, fmt.Errorf("invalid character")
-		}
+	if string(a[0]) != MainnetPrefix && string(a[0]) != TestnetPrefix {
+		return Undef, errors.New("invalid network prefix")
+	}
 
-		dataBytes = append(dataBytes, byte(Base32CharsetReverse[int(b)]))
+	var protocol Protocol
+	switch a[1] {
+	case 48:
+		protocol = ID
+	case 49:
+		protocol = SECP256K1
+	case 50:
+		protocol = Actor
+	case 51:
+		protocol = BLS
+	default:
+		return Undef, errors.New("invalid protocol")
 	}
 
-	if !verifyChecksum(hrp, dataBytes) {
-		return "", 0, nil, fmt.Errorf("invalid checksum")
+	raw := a[2:]
+	if protocol == ID {
+		id, err := strconv.ParseUint(raw, 10, 64)
+		if err != nil {
+			return Undef, errors.New("invalid ID payload")
+		}
+		return newAddress(protocol, leb128.FromUInt64(id)), nil
```

Comment:
we probably shouldnt leb128 encode the bytes form for IDs. Its just taking up extra space (1 bit per byte) and doesnt actually get us anything. Addresses are always delimited in some way anyways. We should just use a standard little endian encoding: https://golang.org/pkg/encoding/binary/#pkg-variables

I realize I didnt catch this in the spec. I'll submit a PR to address this once we have some consensus here.

cc @dignifiedquire

Reply:
I am not sure I agree, all other address formats have a fixed size, except the id version. Not having varints in there could make it tricky to nicely pack addresses in some places I would imagine.

Reply:
For example, with the current impl I can give you a byte array of n addresses, without specifying anything else, and I can parse it in exactly one way (matching protocols, and then determining the length from that) removing the varints this ability gets lost

Reply:
@dignifiedquire True, half of me says "we would never do that", the other half says "but we could".

Reply:
Alright, I'm convinced. It makes sense to me to keep leb128 here